nsenter: add --env for allowing environment variables inheritance

u2386 opened the pull request:
This commit adds support for the -e or --env option in nsenter, allowing a new process to inherit the environment variables from the target process.

If the option is not given, the environment variables will stay the same as in the current namespace.

Example:

    Setup the namespace:
        $ docker run -d -e PROJECT='util linux' --rm alpine sleep 10000
        cb0b69aa7aec
        $ docker inspect --format '{{ .State.Pid }}' cb0b69aa7aec
        470012

    Enter the namespace:
        $ nsenter --all -t 470012 --env env
        PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
        HOSTNAME=cb0b69aa7aec
        PROJECT=util linux
        HOME=/root

u2386 commented:
Close #2080

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +189,56 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+		}
+		bytes_read = read(environ_fd, buffer + size, 4096);
+		if (bytes_read == -1) {
+			err(EXIT_FAILURE, _("read file failed"));
```
err() will exit the process, no need for cleanup/return

u2386 replied:
Thank you for pointing it out.
Updated

t-8ch commented on .vscode/settings.json:
```diff
@@ -0,0 +1,9 @@
+{
```
Drop this file

u2386 replied:
Sorry, I shouldn't have run `git add . `.
Updated.

t-8ch commented on po/ko.po:
```diff
@@ -7,7 +7,7 @@ msgid ""
 "Project-Id-Version: util-linux 2.38-rc4\n"
 "Report-Msgid-Bugs-To: util-linux@vger.kernel.org\n"
-"POT-Creation-Date: 2022-03-16 18:08+0100\n"
+"POT-Creation-Date: 2022-03-28 13:24+0200\n"
```
Drop all changes to .po files

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -510,6 +552,30 @@ int main(int argc, char *argv[])
+	/* Pass environment variables of the target process to the spawned process */
+	if (environ_fd >= 0) {
+		envp = get_process_environment();
+		if (envp == NULL) {
```
No braces around single statements

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -510,6 +552,30 @@ int main(int argc, char *argv[])
+
+		// Set the environment variables from the target process
+		for (int i = 0; envp[i] != NULL; i++) {
+			char* var = envp[i];
```
`char *var` instead of `char* var`

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +191,37 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+	close(environ_fd);
+
+    // Split the buffer into null-terminated strings
+    num = 0;
```
weird indentation

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +191,37 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+	char** envp;
+	int num;
+
+	buffer = (char*)xrealloc(buffer, size);
```
`char *`

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +191,37 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+	char** envp;
+	int num;
+
+	buffer = (char*)xrealloc(buffer, size);
```
The loop is gone, so the `xrealloc()` is just a plain `alloc()`.
It would probably best to introduce a `read_all_alloc()` helper that takes care of reading a file into an allocated buffer.

karelzak replied:
We have ul_path_read_string(), but it reads into buf[BUFSIZ] and strdup() the result. Maybe something more elegant should be implemented.

u2386 replied:
> We have ul_path_read_string(), but it reads into buf[BUFSIZ] and strdup() the result. Maybe something more elegant should be implemented.

Thank you for your suggestion. 
However, using `ul_path_read_string()` is not suitable as it requires a `path_cxt` which is not available in this case. Passing NULL may not be the best practice, although, it works as `ul_path_read_string()` calls `ul_path_open()` which calls `open()` if the path_cxt is NULL.

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +191,37 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+	buffer = (char*)xrealloc(buffer, size);
+	if ((rc = read_all(environ_fd, buffer, size)) < 0)
+		err(EXIT_FAILURE, _("read file failed"));
+	buffer[rc] = '\0';
```
if `rc == 8192` this will write past the allocated memory

t-8ch commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +191,37 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+        p += strlen(p) + 1;
+    }
+
+	envp = (char**)xcalloc(num + 1, sizeof(char*));
```
`char *` and `char **`

u2386 commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +189,56 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+		p++;
+	}
+
+	envp = (char**)malloc((num + 1) * sizeof(char*));
```
Updated.

u2386 commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +189,56 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+	int num;
+
+	do {
+		buffer = (char*)realloc(buffer, size + 4096);
```
Updated.

u2386 commented on sys-utils/nsenter.c:
```diff
@@ -187,6 +189,56 @@ static int is_usable_namespace(pid_t target, const struct namespace_file *nsfile
+			close(environ_fd);
+			return NULL;
+		}
+		bytes_read = read(environ_fd, buffer + size, 4096);
```
Updated.

karelzak commented:
There is lib/env.c with `struct ul_env_list`. Now the code is based on environ[], but it should be simple to add there env_from_fd() to get `struct ul_env_list` from a file-descriptor, and then we can call env_list_setenv() to set all by setenv().

I prefer this solution than implement something nsenter specific.

u2386 commented:
@karelzak  Thanks for the advice. I can work on this these days.

u2386 commented:
> There is `lib/env.c` with `struct ul_env_list`. Now the code is based on environ[], but it should be simple to add there `env_from_fd()` to get `struct ul_env_list` from a file descriptor, and then we can call `env_list_setenv()` to set all by setenv().
> 
> I prefer this solution rather than implementing something nsenter specific.

@karelzak Thank you for the suggestion. I have implemented `env_from_fd()` in env.c and modified the code to use it.

this PR is ready for review

karelzak commented on lib/env.c:
```diff
@@ -81,6 +82,29 @@ static struct ul_env_list *env_list_add(struct ul_env_list *ls0, const char *str
+	struct ul_env_list *ls = NULL;
+
+	if ((rc = read_all_alloc(fd, &buf)) == 0)
+		err(EXIT_FAILURE, _("read of environment failed"));
```
The same problem, do not use err() here. Let's return NULL or so on error.

karelzak commented on term-utils/agetty.c:
```diff
@@ -1507,23 +1508,6 @@ static void auto_baud(struct termios *tp)
 	tcsetattr(STDIN_FILENO, TCSANOW, tp);
 }
 
-static char *xgethostname(void)
```
Please, don't add agetty changes to this PR :-) BTW, agetty uses local xgethostname() due to log_err(); this function writes to syslog.

u2386 replied:
The changes were made because I included "xmalloc" in "all-io.h", which caused a duplicate declaration error during compilation. However, I have now reverted those changes as they were not appropriate.

Thank you for providing the background information.

karelzak commented:
It looks good and more elegant (IMHO) :-)

Please, fix the small issues from the review. Thanks!

karelzak commented on include/all-io.h:
```diff
@@ -18,6 +18,7 @@
 #endif
 
 #include "c.h"
+#include "xalloc.h"
```
And this is also bad in another way, some tools define XALLOC_EXIT_CODE before `include xalloc.h` if you include the file in another header, you will see collisions:
https://download.copr.fedorainfracloud.org/results/packit/util-linux-util-linux-2081/fedora-rawhide-x86_64/05582964-util-linux/builder-live.log.gz

xalloc.h should not be used in lib/  (hmm ... for example lib/logindefs.c uses it, seems like a bad idea, for example for agetty. Seems like an opportunity for another PR :-)

karelzak commented:
Thanks!
